Accepting a type T as options to support custom rendered options

dgeyer-mdsol opened the pull request:
Accept any type as options in react-select.
The types added in 
https://github.com/flow-typed/flow-typed/pull/2817
break the implementations where a custom optionRenderer and filterOptions are passed

pascalduez commented:
Hi @dgeyer-mdsol,

could you please provide samples reproducing the error, or unit tests, or links to the documentation?   
I don't think using `any` (no typechecking) could be an improvement.   

Actually in #2817 I only unclocked the definitions for Flow >= 0.53 
Granted I had the feeling those definitions are not in great shape.

dgeyer-mdsol commented:
@pascalduez Thanks for reviewing, I have added some tests and changed the OtionType to be Object, to support custom rendering and filtering.
This is a link to the docs https://react-select.com/props#prop-types

AndrewSouthpaw commented:
LGTM, @pascalduez it sounds like you have more context, once you give a thumbs up I'll merge it in.

pascalduez commented on definitions/npm/react-select_v1.x.x/flow_v0.53.x-/react-select_v1.x.x.js:
```diff
@@ -1,13 +1,6 @@
-  declare type Option = {
-    [key: string]: any,
-  };
+  declare type OptionType = Object
```
But between `Object` and `{ [key: string]: any }` there's not much differences? 
Except if one need fancy keys.

dgeyer-mdsol replied:
Good observation, actually, I have a case when { [key: string]: any } fails, when options are a Union type. I'll modify the test

pascalduez commented:
Please note that it's the v1 definitions, so the doc is at https://github.com/JedWatson/react-select/tree/v1.x Although I couldn't find more precise refinements.

dgeyer-mdsol commented:
@pascalduez @AndrewSouthpaw I have modified the test so it fails when OptionType is  `{[key: string]: any}` Please have a look and let me know any feedback

pascalduez commented:
It still looks a bit weird to me, especially the union error.   
But should be fine since `Object` is really permissive.   
Maybe @villesau could have a look to get a third eye :)

AndrewSouthpaw commented:
Sounds good, thanks for the in depth review @pascalduez.
